File: c++/Solution347.cpp
//Heap O(nlogk) O(n), n = nums.size()
class Solution347 {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> cnt;
        for(int num : nums)
            ++cnt[num];

        //小排到大
        auto comp = [](pair<int, int>& a, pair<int, int>& b){
                return a.second > b.second;
            };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)> pq(comp);
        for(auto it : cnt){
            pq.push({it.first, it.second});
            if(pq.size() > k)
                pq.pop();
        }

        vector<int> res;
        while(!pq.empty()){
            res.push_back(pq.top().first);
            pq.pop();
        }

        return res;
    }
};

File: java/Solution347.java
package leetCode.java;

import java.util.*;

//Heap O(nlogk) O(n), n = nums.length
class Solution347 {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> counts = new HashMap<>(); // <num, 出現次數>
        for (int num : nums) {
            counts.put(num, counts.getOrDefault(num, 0) + 1);
        }

        // 依照出現次數少到多
        PriorityQueue<Integer> pq = new PriorityQueue((a, b) -> counts.get(a) - counts.get(b));
        for (int num : counts.keySet()) {
            pq.offer(num);
            if (pq.size() > k) {
                pq.poll(); // 移除最少的
            }
        }

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = pq.poll();
        }
        return res;
    }
}